src/VisualStudio/Core/Def/LanguageService/PackageRegistrationTasks.cs

```diff
@@ -0,0 +1,72 @@
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService;
+
+internal sealed class PackageRegistrationTasks(JoinableTaskFactory jtf)
+{
+    private readonly List<Func<IProgress<ServiceProgressData>, PackageRegistrationTasks, CancellationToken, Task>> _backgroundThreadWorkTasks = new();
+    private readonly List<Func<IProgress<ServiceProgressData>, PackageRegistrationTasks, CancellationToken, Task>> _mainThreadWorkTasks = new();
+    private readonly JoinableTaskFactory _jtf = jtf;
+    private readonly object _gate = new();
+
+    public void AddTask(bool isMainThreadTask, Func<IProgress<ServiceProgressData>, PackageRegistrationTasks, CancellationToken, Task> task)
+    {
+        // This lock is a bit extraneous, as the current code doesn't execute any registration of tasks or task processing concurrently.
+        lock (_gate)
+        {
+            if (isMainThreadTask)
+                _mainThreadWorkTasks.Add(task);
+            else
+                _backgroundThreadWorkTasks.Add(task);
+        }
+    }
+
+    public async Task ProcessTasksAsync(IProgress<ServiceProgressData> progress, CancellationToken cancellationToken)
+    {
+        // prime the pump by doing the first group of bg thread work if the initiating thread is not the main thread
+        if (!_jtf.Context.IsOnMainThread)
+            await PerformWorkAsync(useMainThread: false, progress, cancellationToken).ConfigureAwait(false);
+
+        // Continue processing work until everything is completed, switching between main and bg threads as needed.
+        while (_mainThreadWorkTasks.Count > 0 || _backgroundThreadWorkTasks.Count > 0)
+        {
+            await PerformWorkAsync(useMainThread: true, progress, cancellationToken).ConfigureAwait(false);
+            await PerformWorkAsync(useMainThread: false, progress, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    private async Task PerformWorkAsync(bool useMainThread, IProgress<ServiceProgressData> progress, CancellationToken cancellationToken)
+    {
+        var workTasks = useMainThread ? _mainThreadWorkTasks : _backgroundThreadWorkTasks;
+
+        if (workTasks.Count == 0)
+            return;
+
+        // Ensure we're invoking the task on the right thread
+        if (useMainThread)
+            await _jtf.SwitchToMainThreadAsync(cancellationToken);
+        else if (_jtf.Context.IsOnMainThread)
+            await TaskScheduler.Default;
+
+        for (var i = 0; i < workTasks.Count; i++)
+        {
+            var work = workTasks[i];
+
+            // CA(true) is important here, as we want to ensure that each iteration is done in the same
+            // captured context. Thus, even poorly behaving tasks (ie, those that do their own thread switching)
+            // don't effect the next loop iteration.
+            await work(progress, this, cancellationToken).ConfigureAwait(true);
+        }
+
+        workTasks.Clear();
```

Comment:
this is all only ok assumign no concurrency (like running bg tasks alongside main thread tasks).  instead of this, consider locking, copying hte work out, and clearing the copied section.  Note: you could use queus if you're worried about moving costs and what not.

Reply:
Definitely something I've considered and will give a go, just easing my way into these changes.

Reply:
Done! I think this will get changed a bit in the future if I do add concurrency to these lists.